Add share and share array inversion

Drake-Eidukas opened the pull request:
This PR adds in the functionality to invert a Share using the `~` operator. I'm not sure if this is the best way to implement this (as it's somewhat not exactly the definition of inverting that `__invert__` normally targets), but this will make the code for my jubjub application easier.

Drake-Eidukas committed:
Adds share and share array inversion and division

This adds in a pair of mixins for share and share array
inversion, which are used to implement the division
operator for shares and share arrays

Drake-Eidukas commented:
Not sure if `test_mpc.py` is a good place to test the actual inversion operator, but the test is going to be somewhat redundant compared to the ones in `test_mixins.py`.

amiller commented:
Should this use divide instead of invert for this? Not sure if there are guidelines that ~ should be something other than multiplicative inverse. Otherwise looks fine

amiller reviewed (approved):
Looks good

Drake-Eidukas commented:
We should decide if we want to merge this as is so that we can get some existing PRs merged before we achieve git-hell, or if we should first fix #234

smkuls commented:
@Drake-Eidukas can you rebase this once with `dev`. Lets merge this first and then take up #234.

smkuls commented on honeybadgermpc/mpc.py:
```diff
@@ -239,6 +239,23 @@ def __mul__(self, other):
         def __str__(self): return '{%d}' % (self.v)
 
+        def __div__(self, other):
+            if MixinOpName.InvertShare not in context.config:
```
I am wondering if we should write a MixinValidator, now that we have started creating dependent mixins this might end up with a lot of if, else checks. What do you think @Drake-Eidukas ?

Drake-Eidukas replied:
I think a validator might not be too impossible-- especially if we are considering redoing some of the way we do mixins (i.e. moving stuff into progs). Could be interesting to create some way to have the mixins define the operations they need. I'll think about this some more when I start tackling the refactors to mixins

smkuls commented on tests/test_mixins.py:
```diff
@@ -182,3 +164,52 @@
+
+
+@mark.asyncio
+async def test_invert_share(test_preprocessing):
```
Do you want to consider adding a negative test case where you don't pass the mixin and check for `NotImplementedError`?

Drake-Eidukas replied:
I'll add this-- good catch

smkuls commented on tests/test_mixins.py:
```diff
@@ -182,3 +164,52 @@
+    async def _prog(context):
+        share = test_preprocessing.elements.get_rand(context)
+        # assert 1/x * x == 1
+        assert await(await(await(Inverter.invert_share(context, share))
```
We need to fix this. This is crazy. This needs to align with the data flow design. Lets address this after this PR.

Drake-Eidukas replied:
Agreed-- it was a massive PITA to figure out how to get this to compile and all... err, run without crashing

Drake-Eidukas commented:
@smkuls Should be good to go, just rebased this and added that test. I also moved the tests for divide into test_mixins to fit where multiplication is. Let me know if you want to have that moved back.

smkuls reviewed (approved):
LGTM!
